provide.go

```diff
@@ -72,14 +72,45 @@ type provideOption struct {
 }
 
 func (o provideOption) apply(mod *module) {
+	private := false
+
+	var targets []interface{}
 	for _, target := range o.Targets {
+		if _, ok := target.(privateProvideOption); ok {
+			private = true
+			continue
+		}
+		targets = append(targets, target)
+	}
+
+	for _, target := range targets {
 		mod.provides = append(mod.provides, provide{
-			Target: target,
-			Stack:  o.Stack,
+			Target:  target,
+			Stack:   o.Stack,
+			Private: private,
 		})
 	}
 }
 
+type privateProvideOption struct{}
+
+// Private returns an option that can be passed as an argument to [Provide]
+// to restrict access to the constructors being provided. Specifically,
+// corresponding constructors can only be used within the current module
+// or modules the current module contains. Other modules that contain this
+// module won't be able to use the constructor.
+//
+// For example, the following would fail because the app doesn't have access
+// to the inner module's constructor.
+//
+//	fx.New(
+//		fx.Module("SubModule", fx.Provide(func() int { return 0 }, fx.Private())),
+//		fx.Invoke(func(a int) {}),
+//	)
+func Private() privateProvideOption {
```

Comment:
This seems like a linter warning? Should we make a public interface for this to resolve it? I'm not sure what's the best route here.

Reply:
Yeah, I would declare a new interface and seal it with an unexported method. Then make privateProvideOption implement that. Maybe 'ProvideOption'?

```go
type ProvideOption interface{ provideOption() }
func Private() ProvideOption { ... }
```

---

Separately, consider making `func Private(..)` accept a boolean:

```go
fx.Provide(a, b, c, fx.Private(true))
```

----

Another possibility is to make `fx.Private` a top-level variable with an unexported type.

```go
var Private = privateOption{}
```

Usage: `fx.Provide(..., fx.Private)`.

That way, you can mess with the type later like add new methods, make it implement other interfaces (like if Private needs to be a valid `Option` for something) without committing to it being a provide-specific thing.

Reply:
Thx for the ideas. I opted for the top-level var for the reasons you mentioned and to avoid awkward naming conflicts (there already exists a struct called provideOption)